### PR TITLE
User/diegoarmstrong/vds/data processing

### PR DESCRIPTION
**Pull request by DiegoArmstrong**
Added functionality to process CAN messages from the ADC readings for vds sensors.

### CODE REVIEW

File: components/vds/Core/Inc/can.h [hunk 1]
```diff
@@ -27,6 +27,7 @@ extern "C" {
 
 /* Includes ------------------------------------------------------------------*/
 #include "main.h"
+#include "common.h"
```
Comment (ishanjoshi23) [inline components/vds/Core/Inc/can.h]: This needs to be moved below to the `USER CODE BEGIN Includes` section. The autogenerator will remove this where it is right now.

File: components/vds/Core/Inc/can.h [hunk 2]
```diff
@@ -38,13 +39,31 @@ extern CAN_HandleTypeDef hcan2;
...
+
+
 /* USER CODE END Private defines */
 
 void MX_CAN1_Init(void);
 void MX_CAN2_Init(void);
 
 /* USER CODE BEGIN Prototypes */
-
+void CAN_SendShockTravel(VDS_Data_t *adc_data);
```
Comment (ishanjoshi23) [inline components/vds/Core/Inc/can.h:63]: Good naming convention!

File: components/vds/Core/Src/adc.c [hunk 3]
```diff
@@ -128,6 +128,10 @@ void MX_ADC1_Init(void)
   }
   /* USER CODE BEGIN ADC1_Init 2 */
 
+  if (HAL_ADCEx_Calibration_Start(&hadc1) != HAL_OK){
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/adc.c]: I don't think we should be including this in production code, as recalibrating repeatedly would impact the characterization functions. @mjohal67 Please correct me if I am wrong.

File: components/vds/Core/Src/can.c [hunk 4]
```diff
@@ -22,6 +22,93 @@
...
+
+void CAN_SendShockTravel(VDS_Data_t *vds_data)
+{
+    CAN_Message_t message;
+    message.header = shock_travel_header;
+
+    ADC_Value_t adc_value;
+
+    adc_value.value = vds_data->adc_data.ADC_shock_travel_1;
+    message.data[0] = adc_value.bytes.low;
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/can.c]: All the `message.data[i]` can be combined into a for loop to reduce the code necessary for this.

File: components/vds/Core/Src/can.c [hunk 5]
```diff
@@ -22,6 +22,93 @@
...
+
+void CAN_SendBrakeAndSteering(VDS_Data_t *vds_data)
+{
+    CAN_Message_t message;
+    message.header = brake_and_steering_header;
+
+    ADC_Value_t adc_value;
+
+    adc_value.value = vds_data->adc_data.ADC_brake_pressure_1;
+    message.data[0] = adc_value.bytes.low;
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/can.c]: All the `message.data[i]` can be combined into a for loop to reduce the code necessary for this.

File: components/vds/Core/Src/can.c [hunk 6]
```diff
@@ -22,6 +22,93 @@
...
+    message.data[5] = adc_value.bytes.high;
+
+    adc_value.value = vds_data->adc_data.ADC_steering_angle;
+    message.data[6] = adc_value.bytes.low;
+    message.data[7] = adc_value.bytes.high;
+
+    HAL_CAN_AddTxMessage(&hcan1, &message.header, message.data, &can_mailbox);
+}
+
+void CAN_SendVDSDiagnostic(VDS_Data_t *vds_status){
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/can.c]: @AarjavJain101 Will discuss the new diagnostic template with you soon.

File: components/vds/Core/Src/main.c [hunk 7]
```diff
@@ -260,6 +269,48 @@ void HAL_ADC_ErrorCallback(ADC_HandleTypeDef *hadc)
   }
 }
 
+void convertADCValue(void){
+
+  vds_data.adc_data.ADC_brake_pressure_1 = (24.55 * vds_data.adc_data.ADC_brake_pressure_1) + 1.925;
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/main.c]: No magic numbers. Please define these as symbolic constants and name them appropriately.

File: components/vds/Core/Src/main.c [hunk 8]
```diff
@@ -260,6 +269,48 @@ void HAL_ADC_ErrorCallback(ADC_HandleTypeDef *hadc)
   }
 }
 
+void convertADCValue(void){
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/main.c]: Please move helper functions to their own file. E.g., ADC helpers can be moved to the USER code part of the `adc.c` file. Such functions should be named like ADC_convert_ADC_Value where the function starts with the peripheral/file it lives in.

File: components/vds/Core/Src/main.c [hunk 9]
```diff
@@ -260,6 +269,48 @@ void HAL_ADC_ErrorCallback(ADC_HandleTypeDef *hadc)
...
+  vds_data.adc_data.ADC_shock_travel_1 = (0.1595 * vds_data.adc_data.ADC_shock_travel_1) + 2.3875;
+  vds_data.adc_data.ADC_shock_travel_2 = (0.1595 * vds_data.adc_data.ADC_shock_travel_2) + 2.3875;
+  vds_data.adc_data.ADC_shock_travel_3 = (0.1595 * vds_data.adc_data.ADC_shock_travel_3) + 2.3875;
+  vds_data.adc_data.ADC_shock_travel_4 = (0.1595 * vds_data.adc_data.ADC_shock_travel_4) + 2.3875;
+
+  // TODO: Add conversion for steering angle once sensor has been selected and tested
+
+}
+
+void processCANMessages(void)
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/main.c]: Same for this function. Would be good if it was in can.c as CAN_process_CAN_Messages

File: components/vds/Core/Src/main.c [hunk 10]
```diff
@@ -260,6 +269,48 @@ void HAL_ADC_ErrorCallback(ADC_HandleTypeDef *hadc)
...
+{
+  uint32_t current_time = HAL_GetTick();
+
+
+  // Send shock travel message at 1kHz
+  if (current_time - last_shock_travel_time >= 1)
+  {
+    CAN_SendShockTravel(&vds_data);
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/main.c]: Just like this! This is good naming.

File: components/vds/Core/Src/main.c [hunk 11]
```diff
@@ -260,6 +269,48 @@ void HAL_ADC_ErrorCallback(ADC_HandleTypeDef *hadc)
...
+  }
+
+  // Send brake and steering message at 10Hz
+  if (current_time - last_brake_and_steering_time >= 100)
+  {
+    CAN_SendBrakeAndSteering(&vds_data);
+    last_brake_and_steering_time = current_time;
+  }
+
+  // Send diagnostic message every 5 seconds
```
Comment (ishanjoshi23) [inline components/vds/Core/Src/main.c]: @AarjavJain101 Will discuss this on Saturday

**Review by ishanjoshi23 — CHANGES_REQUESTED** [review]

**Comment by ishanjoshi23** [comment]
Need  to add/review  averaging and add IWDG @makac1896

File: components/vds/Core/Inc/can.h (hunk 2 above)
Comment (DiegoArmstrong) [reply components/vds/Core/Inc/can.h:63]: Thanks bbg

**Comment by ishanjoshi23** [comment]
Changes look good. Merging into main feature branch

**Pull request merged**